### PR TITLE
ovirt-image-template: add glance as disk source

### PR DESCRIPTION
**Pull request by tareqalayan**
Before this you could create a template from a qcow file located in some
webserver.
Now you can also create a template froma disk that is located in glance.

see:
https://github.com/oVirt/ovirt-ansible/issues/29

### CODE REVIEW

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by tareqalayan** [comment]
@machacekondra 
do you think it is a good idea to enhance this role?
instead we can have a different role that adds the provider and exports the disk? 

aren't roles meant to do one thing ?

i think it would be easier for the user and for the maintainer? what do u think ?

**Comment by mwperina** [comment]
ci test please

**Comment by machacekondra** [comment]
I thought it will just create a template from a disk in a glance. Not add a glance, it doesn't belong to this role.

**Comment by tareqalayan** [comment]
@machacekondra so should i remove the part of adding the glance. And assume that a glance is connected?

**Comment by machacekondra** [comment]
Yes, adding a glance (or any other external provider) should be a different role IMO.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

File: roles/ovirt-image-template/README.md [hunk 1]
```diff
@@ -1,7 +1,9 @@
 oVirt Image Template
 ====================
 
-The `ovirt-image-template` role downloads a QCOW2 image from a specified URL and creates a template from it.
```
Comment (machacekondra) [inline roles/ovirt-image-template/README.md:4]: I would rephrase it to:

The `ovirt-image-template` role creates a template from external image. Currently the disk can be a image in Glance external provider or QCOW2 image.

File: roles/ovirt-image-template/README.md [hunk 2]
```diff
@@ -15,7 +17,8 @@ Role Variables
 --------------
 
 | Name               | Default value         |                            |
-|--------------------|-----------------------|----------------------------| 
+|--------------------|-----------------------|----------------------------|
+| source             | qcow_file             | The source to create the template from. Source can be qcow_file or glance |
```
Comment (machacekondra) [inline roles/ovirt-image-template/README.md]: I think this isn't really needed.

File: roles/ovirt-image-template/README.md [hunk 3]
```diff
@@ -28,6 +31,18 @@ Role Variables
 | template_disk_interface | virtio           | Interface of the template disk. |
 | template_timeout   | 600                   | Amount of time to wait for the template to be created. |
 | template_nics      | {name: nic1, profile_name: ovirtmgmt, interface: virtio} | List of dictionaries that specify the NICs of template. |
+| external_templates   | UNDEF               | List of dictionaries that describes the templates |
```
Comment (machacekondra) [inline roles/ovirt-image-template/README.md]: Not sure what's this for?

File: roles/ovirt-image-template/meta/main.yml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 galaxy_info:
   author: Ondra Machacek
-  description: Role to create template from qcow2 image.
+  description: Role to create template from qcow2 image or disk resides in glance.
```
Comment (machacekondra) [inline roles/ovirt-image-template/meta/main.yml]: Role to create template from external image.

File: roles/ovirt-image-template/tasks/main.yml [hunk 5]
```diff
@@ -1,23 +1,31 @@
 ---
+- name: source of the disk to import
```
Comment (machacekondra) [inline roles/ovirt-image-template/tasks/main.yml]: Please remove.

File: roles/ovirt-image-template/tasks/main.yml [hunk 6]
```diff
@@ -1,23 +1,31 @@
...
+  debug:
+    var: source
+
 - name: Download the qcow image
   get_url:
     url: "{{ qcow_url }}"
     dest: "{{ image_path }}"
   register: downloaded
+  when: source  == "qcow_file"
```
Comment (machacekondra) [inline roles/ovirt-image-template/tasks/main.yml]: Please separate it into two files:
- qcow2_image.yml
- glance_image.yml

and import it based on params, something like:

```yaml
- name: Fail when invalid parameters
  fail:
    msg: "You must either specify qcow2_url or glance_image"
  when: "glance_image is defined and qcow2_url is defined"

- name: Glance image
  include: glance_image.yml
  when: "glance_image is defined"

- name: QCOW2 image
  include: qcow2_image.yml
  when: "qcow2_url is defined"
```

**Review by machacekondra — CHANGES_REQUESTED** [review]

File: roles/ovirt-image-template/README.md (hunk 1 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/README.md:4]: done

File: roles/ovirt-image-template/README.md (hunk 2 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: done

File: roles/ovirt-image-template/README.md (hunk 3 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: templates to be imported from the glance.

File: roles/ovirt-image-template/meta/main.yml (hunk 4 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/meta/main.yml]: done

File: roles/ovirt-image-template/tasks/main.yml (hunk 5 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/tasks/main.yml]: done

File: roles/ovirt-image-template/tasks/main.yml (hunk 6 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/tasks/main.yml]: done

**Review by tareqalayan — COMMENTED** [review]
done according to your comments

File: roles/ovirt-image-template/README.md (hunk 3 above)
Comment (machacekondra) [reply roles/ovirt-image-template/README.md]: There are no templates in glance, that's just images. And for template we already have variables.

Comment (machacekondra) [reply roles/ovirt-image-template/README.md]: We don't need that we already have eveything defined, the only two new parameters should be `glance_image` and `glance_image_provider`.

File: roles/ovirt-image-template/README.md [hunk 7]
```diff
@@ -58,8 +70,21 @@ Example Playbook
...
+        cluster: cluster_1
+      - name: mytemplate_73
+        image_provider: qe-infra-glance
+        image_disk: rhel7.3_ovirt4.2_guest_disk
+        storage_domain: nfs_0
+        cluster: cluster_1
+
   roles:
-    - ovirt-image-template
+    - { role: ovirt-image-template, source: 'qcow_file' }
```
Comment (machacekondra) [inline roles/ovirt-image-template/README.md]: please change back.

Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: done

File: roles/ovirt-image-template/README.md (hunk 3 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: I want to have the ability to import more than one template. 
so this role accepts a list of external_templates from provider and imports them as templates ..

Comment (machacekondra) [reply roles/ovirt-image-template/README.md]: You can execute role multiple times to achieve that. But if you insist, then this isn't a material for this role.

Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: If you have a list it will be more scalable, you can have list of one disk or many, so u cover both cases.
running same role multiple times is maybe good when it done manually. 
we want to use these roles for automating things. 
I really suggest to change this role to also accept a list of qcow images ... 

i am not sure if looping over roles in the playbook is supported in the ansible.

Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: @machacekondra  what do you think?

Comment (machacekondra) [reply roles/ovirt-image-template/README.md]: I think `with_items` works with `include_role` so it should work. We can add it, but most of the roles work with just one item, so our roles do this as well. Or you can use multiple plays.

Comment (machacekondra) [reply roles/ovirt-image-template/README.md]: I don't like this `external_templates` var. This isn't consistent. This looping should be done on Ansible side: https://github.com/ansible/ansible/issues/24921 someday in future. I think it brings the advatange of asynchronousness for example and simplifies the role.

So if you need that role, can you just create you own role, which won't be part of this repo, so we don't need to support it, and let's add it once users need it? Users still can create what they want in post/pre tasks, those roles just wraps module into role, so it don't do anything basically. Let's talk about it on Monday meeting.

**Comment by machacekondra** [comment]
One possible way of having what you need is:

```yaml
---
#######################
# Play to setup datacenter
- hosts: localhost
  connection: local

  roles:
    - ovirt-infra

#######################
# Play to setup tempaltes
- hosts: localhost
  connection: local

  vars:
    images:
      - qcow_url: http://test1.com/image.qcow2
      - qcow_url: http://test2.com/image.qcow2

  tasks:
    - include_role:
        name: "ovirt-image-template"
      vars:
        qcow_url: "{{ item.qcow_url }}" 
      with_items: "{{ images }}"

#######################
# Play to setup VM infra
- hosts: localhost
  connection: local

  roles:
    - ovirt-vm-infra

```

Is that OK?

In future it should be something like:

```yaml
- hosts: localhost
  connection: local
  
  images:
    - qcow_url: 'http://abc/image1'
    - qcow_url: 'http://abc/image2'

  roles:
     - { 'name': 'ovirt-infra' }
     - { 'name': 'ovirt-image-template': with_items: "{{ images }}" }
     - { 'name': 'ovirt-vm-infra' }
```

File: roles/ovirt-image-template/README.md [hunk 8]
```diff
@@ -58,6 +60,10 @@ Example Playbook
     template_disk_size: 10GiB
     template_disk_storage: mydata
 
+    # if you want to import image from glance use these vars and remove teh qcow_url
```
Comment (machacekondra) [inline roles/ovirt-image-template/README.md]: Better to create a new example.

File: roles/ovirt-image-template/tasks/glance_image.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  - name: Import templates from glance
+    ovirt_templates:
+      auth: "{{ ovirt_auth }}"
+      state: imported
+      name: "{{ template_name }}"
+      image_provider: "{{ glance_image_provider }}"
+      image_disk: "{{ glance_image }}"
+      storage_domain: "{{ template_disk_storage | default(disk_storage_domain.name) }}"
+      cluster: "{{ template_cluster }}"
+      fetch_nested: True
```
Comment (machacekondra) [inline roles/ovirt-image-template/tasks/glance_image.yml]: Line 38 and 39 is needed why?

File: roles/ovirt-image-template/README.md [hunk 10]
```diff
@@ -28,6 +28,8 @@ Role Variables
 | template_disk_interface | virtio           | Interface of the template disk. |
 | template_timeout   | 600                   | Amount of time to wait for the template to be created. |
 | template_nics      | {name: nic1, profile_name: ovirtmgmt, interface: virtio} | List of dictionaries that specify the NICs of template. |
+| glance_image_provider        | UNDEF            | Name of the glance image provider.                    |
+| glance_image            | UNDEF               | this parameter specifies the name of disk to be imported as template. |
```
Comment (machacekondra) [inline roles/ovirt-image-template/README.md]: This parameter specifies the name of disk in glance provider to be imported as template.

Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: done

File: roles/ovirt-image-template/README.md (hunk 8 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/README.md]: done

Comment (mykaul) [inline roles/ovirt-image-template/README.md]: teh -> the

File: roles/ovirt-image-template/tasks/glance_image.yml (hunk 9 above)
Comment (tareqalayan) [reply roles/ovirt-image-template/tasks/glance_image.yml]: no need , these are leftovers ...

**Review by machacekondra — APPROVED** [review]
